### PR TITLE
Fody tracer- adding entry\exit logs on compilation

### PR DESCRIPTION
**Pull request by noescape00**
Adds additional compilation step that will add trace logging to all classes under Stratis.Bitcoin namespace.

This implements first step from list of steps covered in #2249 

PS:
to review this compile the solution, run it with trace logging on (any target with namespace Stratis.Bitcoin will do)- logs will contain entry and exit logs for every method. 

PSS:
To understand how exactly it works read links supplied in #2249

### CODE REVIEW

**Comment by bokobza** [comment]
do we need to review the Fody to NLog adapter or is this something you brought from somewhere else and that's already tested, used in the wild etc?

**Comment by noescape00** [comment]
>do we need to review the Fody to NLog adapter or is this something you brought from somewhere else and that's already tested, used in the wild etc?

You don't need to, it all came from Fody and Tracer

Here: 
https://github.com/csnemes/tracer/blob/master/Tracer.NLog/Log.cs
https://github.com/csnemes/tracer/blob/master/Tracer.NLog/Adapters/LoggerAdapter.cs
https://github.com/csnemes/tracer/blob/master/Tracer.NLog/Adapters/LogManagerAdapter.cs

But I've modified `LoggerAdapter` to match our coding style and logic there to match logging style

**Comment by noescape00** [comment]
The tests are passing on my PC but it looks like we need to configure CI to work with this change.

**Comment by dangershony** [comment]
Ok looks cool but I need to have a good look at whats going on here.
Do we need to delete all entry and exist logs at some point? (will they become duplicated?)

**Comment by dangershony** [comment]
>But I've modified LoggerAdapter to match our coding style and logic there so it produces logs that match our logging style.

If it does not match our logging style this is no big deal we can change our style a bit if needed.

**Comment by noescape00** [comment]
>Do we need to delete all entry and exist logs at some point? (will they become duplicated?)

After this PR they are already duplicated. 
So very next PR will be about removing those from the whole solution

**Comment by noescape00** [comment]
>If it does not match our logging style this is no big deal we can change our style a bit if needed.

Yeah, I'll play with it so it's comfortable to read. 
But it's already pretty good.

**Comment by noescape00** [comment]
@dangershony @bokobza fixed CI.

File: appveyor.yml [hunk 1]
```diff
@@ -39,6 +39,7 @@ init:
 build_script:
 - ps: |
       cd src
+      dotnet remove Stratis.Bitcoin/Stratis.Bitcoin.csproj package Fody
```
Comment (dangershony) [inline appveyor.yml]: What does this do? we build appvoyar without logs?

File: build.sh [hunk 2]
```diff
@@ -1,7 +1,12 @@
 #!/bin/bash
 dotnet --info
-echo STARTED dotnet build
+
+echo disabling Fody because it is not needed for CI
```
Comment (dangershony) [inline build.sh]: ah ok this answers my question, yea not needed for CI indeed

**Comment by fenix2222** [comment]
Overall looks good. I have used Fody weavers before and they work really well.

File: src/FodyWeavers.xml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<Weavers>
+  <Tracer adapterAssembly="FodyNlogAdapter"
+      logManager="FodyNlogAdapter.Adapters.LogManagerAdapter"      
+      logger="FodyNlogAdapter.Adapters.LoggerAdapter"
+      staticLogger="FodyNlogAdapter.Log"
+      traceConstructors="false"
+      traceProperties="false">
+      <TraceOn namespace="Stratis.Bitcoin+*" class="private" method="private" />
```
[the remark this answers is not in the document]
Comment (noescape00) [reply src/FodyWeavers.xml]: That's probably up to smart contracts guys to decide. 

@codingupastorm would you like to have trace logging enabled for all smart contracts projects?

File: src/FodyNlogAdapter/FodyNlogAdapter.csproj [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="NLog" Version="4.5.10" />
```
[the remark this answers is not in the document]
Comment (noescape00) [reply src/FodyNlogAdapter/FodyNlogAdapter.csproj]: Yeah, Stratis.Bitcoin is targeting alpha version. 
So probably should change it here. 
Good catch

File: src/FodyWeavers.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (codingupastorm) [reply src/FodyWeavers.xml]: Ooh thanks for the ping Ivan.

Hmm at the moment  it's not in NBitcoin, only our features ?

We should probably put it in .SmartContracts.Core and .SmartContracts.Executor.Reflection, will be easier if all node code is consistent

[the remark this answers is not in the document]
Comment (noescape00) [reply src/FodyWeavers.xml]: >Hmm at the moment it's not in NBitcoin, only our features ?

Correct. Only for stuff which namespace starts with `Stratis.Bitcoin`

>We should probably put it in .SmartContracts.Core and .SmartContracts.Executor.Reflection, will be easier if all node code is consistent

Cool. Will add.

**Pull request merged**